File: src/lib/api/logging.ts
import { invoke } from '@tauri-apps/api/core';

export type LogLevel = 'error' | 'warn' | 'info' | 'debug' | 'trace';

/**
 * Send a log entry from the frontend to the backend logging system.
 * This persists the log to the rolling log file alongside Rust-side logs.
 */
export async function frontendLog(
  level: LogLevel,
  message: string,
  context?: Record<string, unknown>,
): Promise<void> {
  return invoke('frontend_log', { level, message, context: context ?? null });
}

/** Return the absolute path to the log directory. */
export async function getLogPath(): Promise<string> {
  return invoke('get_log_path');
}

/** Return the last N lines from the most recent log file. */
export async function getRecentLogs(lines?: number): Promise<string> {
  return invoke('get_recent_logs', { lines: lines ?? 200 });
}

/**
 * Global logger that writes to both console and the backend log file.
 *
 * Usage:
 *   import { log } from '$lib/api/logging';
 *   log.error('Transcription failed', { component: 'RecordTab', recordingId });
 *   log.info('Recording started');
 */
export const log = {
  error(message: string, context?: Record<string, unknown>) {
    console.error(`[FerriScribe] ${message}`, context ?? '');
    frontendLog('error', message, context).catch(() => {});
  },
  warn(message: string, context?: Record<string, unknown>) {
    console.warn(`[FerriScribe] ${message}`, context ?? '');
    frontendLog('warn', message, context).catch(() => {});
  },
  info(message: string, context?: Record<string, unknown>) {
    console.info(`[FerriScribe] ${message}`, context ?? '');
    frontendLog('info', message, context).catch(() => {});
  },
  debug(message: string, context?: Record<string, unknown>) {
    console.debug(`[FerriScribe] ${message}`, context ?? '');
    frontendLog('debug', message, context).catch(() => {});
  },
};

File: src-tauri/src/commands/logging.rs
//! Tauri commands for the logging subsystem.
//!
//! Provides:
//! - `get_log_path`   — returns the log directory so the UI can offer "Open logs"
//! - `get_recent_logs` — returns the tail of the current log file for in-app viewing
//! - `frontend_log`   — bridge for the frontend to write structured log entries

use std::path::PathBuf;

use medical_core::error::{AppError, AppResult};

/// Return the path to the log directory.
#[tauri::command]
pub fn get_log_path() -> String {
    log_dir().display().to_string()
}

/// Return the last `lines` lines of today's log file.
///
/// Useful for an in-app log viewer or for attaching to bug reports.
#[tauri::command]
pub fn get_recent_logs(lines: Option<usize>) -> AppResult<String> {
    let max_lines = lines.unwrap_or(200);
    let dir = log_dir();

    // Find the most recently modified .log file
    let mut newest: Option<(std::time::SystemTime, PathBuf)> = None;
    let entries = std::fs::read_dir(&dir)?;
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("log") {
            continue;
        }
        if let Ok(meta) = path.metadata() {
            if let Ok(modified) = meta.modified() {
                if newest.as_ref().map_or(true, |(t, _)| modified > *t) {
                    newest = Some((modified, path));
                }
            }
        }
    }

    let log_path = newest
        .map(|(_, p)| p)
        .ok_or_else(|| AppError::Other("No log files found".to_string()))?;

    let content = std::fs::read_to_string(&log_path)?;

    let tail: Vec<&str> = content.lines().rev().take(max_lines).collect();
    let result: Vec<&str> = tail.into_iter().rev().collect();

    // Prepend source file name for context
    let filename = log_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown");
    Ok(format!("--- {filename} (last {max_lines} lines) ---\n{}", result.join("\n")))
}

/// Bridge for frontend JavaScript to log structured entries to the backend.
///
/// Call from the frontend as:
/// ```js
/// invoke('frontend_log', { level: 'error', message: 'Something failed', context: { component: 'RecordTab' } })
/// ```
#[tauri::command]
pub fn frontend_log(level: String, message: String, context: Option<serde_json::Value>) {
    let ctx = context
        .as_ref()
        .map(|v| v.to_string())
        .unwrap_or_default();

    match level.to_lowercase().as_str() {
        "error" => tracing::error!(source = "frontend", context = %ctx, "{message}"),
        "warn" => tracing::warn!(source = "frontend", context = %ctx, "{message}"),
        "debug" => tracing::debug!(source = "frontend", context = %ctx, "{message}"),
        "trace" => tracing::trace!(source = "frontend", context = %ctx, "{message}"),
        _ => tracing::info!(source = "frontend", context = %ctx, "{message}"),
    }
}

fn log_dir() -> PathBuf {
    dirs::data_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("rust-medical-assistant")
        .join("logs")
}
